### PR TITLE
Update config to split `jax` and `econpizza` cache

### PR DESCRIPTION
**Pull request by shtopane**
- add property `enable_jax_persistent_cache`
- add function to setup folder
- update tests

### CODE REVIEW

File: econpizza/config.py [hunk 1]
```diff
@@ -19,6 +20,15 @@ def enable_persistent_cache(self, value):
         self._enable_persistent_cache = value
         self.setup_persistent_cache()
 
+    @property
+    def enable_jax_persistent_cache(self):
```
Comment (gboehl) [inline econpizza/config.py]: I don't see why this property is needed.

File: econpizza/config.py [hunk 2]
```diff
@@ -19,6 +20,15 @@ def enable_persistent_cache(self, value):
...
+        return self._enable_jax_persistent_cache
+
+    @enable_jax_persistent_cache.setter
```
Comment (gboehl) [inline econpizza/config.py]: Can be removed once `update` is the only interface.

File: econpizza/config.py [hunk 3]
```diff
@@ -49,8 +59,9 @@ def _create_cache_dir(self, folder_name: str):
         return folder_path
 
     def setup_persistent_cache(self):
```
Comment (gboehl) [inline econpizza/config.py]: This should be included in `update` (or called from there).

File: econpizza/config.py [hunk 4]
```diff
@@ -59,8 +70,12 @@ def setup_persistent_cache(self):
             else:
                 folder_path_pizza = self.econpizza_cache_folder
```
Comment (gboehl) [inline econpizza/config.py]: Any `path_pizza` remainings should for now be removed.

File: econpizza/config.py [hunk 5]
```diff
@@ -59,8 +70,12 @@ def setup_persistent_cache(self):
...
 
-            # Jax cache is enabled by the used via JAX API. In this case we should not set another folder
-            if jax.config.jax_compilation_cache_dir is None:
+    def setup_persistent_cache_jax(self):
```
Comment (gboehl) [inline econpizza/config.py:33]: Should also be included in `update`.

**Review by gboehl — CHANGES_REQUESTED** [review]
Great! So let us now *please* simplify the interface to using `ep.config.update(STUFF)` only. The two `setter` functions can also be removed then.

`update` should allow to change the folder path or to enable/disable caching. It should then check whether the change in attributes should trigger any action (like creating folders or disabling jax cache). Do you understand what I mean?

**Comment by shtopane** [comment]
Here's a proposed version of the code using only `update`. The easier plug-in, plug-out way to remove econpizza cache is to remove the if clause in the update method - keep the folder property, but don't handle the case inside the update.
If this version is acceptable, I'll push it and change the tests.
```python
class EconPizzaConfig(dict):
    def __init__(self, *args, **kwargs):
        super(EconPizzaConfig, self).__init__(*args, **kwargs)
        self._enable_persistent_cache = False
        self._enable_jax_persistent_cache = False
        self._jax_cache_folder = "__jax_cache__"
        self._econpizza_cache_folder = "__econpizza_cache__"

    def update(self, key, value):
        """Updates the attribute, and if it's related to caching, calls the appropriate setup function."""
        if key == "enable_persistent_cache": # <- Remove this if case to remove econpizza cache for now
            self._enable_persistent_cache = value

            if self._enable_persistent_cache:
                self.setup_persistent_cache()
        elif key == "enable_jax_persistent_cache":
            self._enable_jax_persistent_cache = value

            if self._enable_jax_persistent_cache:
                self.setup_persistent_cache_jax()
        elif key == "jax_cache_folder":
            self._jax_cache_folder = value
        elif key == "econpizza_cache_folder":
            self._econpizza_cache_folder = value
        else:
            raise AttributeError(f"'EconPizzaConfig' object has no attribute '{key}'")

    def _create_cache_dir(self, folder_name: str):
        cwd = os.getcwd()
        folder_path = os.path.join(cwd, folder_name)
        os.makedirs(folder_path, exist_ok=True)
        return folder_path

    def setup_persistent_cache(self):
        """Create econpizza cache folder. If caching is enabled, sets up the cache."""
        if not os.path.exists(self._econpizza_cache_folder):
            folder_path_pizza = self._create_cache_dir(self._econpizza_cache_folder)
            self._econpizza_cache_folder = folder_path_pizza
        else:
            folder_path_pizza = self._econpizza_cache_folder

    def setup_persistent_cache_jax(self):
        """Setup JAX persistent cache if enabled."""
        if jax.config.jax_compilation_cache_dir is None and not os.path.exists(self._jax_cache_folder):
            folder_path_jax = self._create_cache_dir(self._jax_cache_folder)
            jax.config.update("jax_compilation_cache_dir", folder_path_jax)
            jax.config.update("jax_persistent_cache_min_entry_size_bytes", -1)
            jax.config.update("jax_persistent_cache_min_compile_time_secs", 0)
            self._jax_cache_folder = folder_path_jax

    def __repr__(self):
        properties = {
            k.lstrip("_"): v for k, v in self.__dict__.items() if k.startswith("_")
        }
        return f"{properties}"

    def __str__(self):
        return self.__repr__()
```

**Comment by gboehl** [comment]
I can't add inline comments here, so please make a commit next time. But yes, seems to be on the right track. I don't see why the class attributes need an underscore. 

Why did you remove the old way `update` worked using `setattr(self, key, value)`? I think that was far more elegant. Your class is a `dict` where keys can either be accessed via `config[key]' or `config.key`, so setting values should be straightforward without the `if`/`else` causes?

**Comment by shtopane** [comment]
@gboehl Please have a look again - hope this version is closer to what you have in mind.

File: econpizza/config.py [hunk 6]
```diff
@@ -1,80 +1,53 @@
...
-    
+
     def _create_cache_dir(self, folder_name: str):
         cwd = os.getcwd()
         folder_path = os.path.join(cwd, folder_name)
         os.makedirs(folder_path, exist_ok=True)
-
         return folder_path
 
     def setup_persistent_cache(self):
```
Comment (gboehl) [inline econpizza/config.py]: Thanks, looking good now! Last thing: this function only does econpizza cache things, which we agreed are for now always disabled. Could you remove this function?

Comment (shtopane) [reply econpizza/config.py]: Certainly! Just to be clear: it's not called right now even if `enable_persistent_cache` is set to `True`.
So, my question is do I remove the function along with the tests for it (which are now skipped) and the flag itself (`enable_persistent_cache`)?

Comment (gboehl) [reply econpizza/config.py]: I would say so. You can put it in the other pull request for later.

Comment (shtopane) [reply econpizza/config.py]: Okay, done. Please have a quick look again and I think we can merge.

Comment (gboehl) [reply econpizza/config.py]: Looks good. The mapping as a dict of functions is a nice idea!

File: econpizza/config.py [hunk 7]
```diff
@@ -1,80 +1,42 @@
...
-    def __str__(self):
-        return self.__repr__()
-
+    def setup_persistent_cache_jax(self):
+        """Setup JAX persistent cache if enabled."""
+        if jax.config.jax_compilation_cache_dir is None and not os.path.exists(self.jax_cache_folder):
+            folder_path_jax = self._create_cache_dir(self.jax_cache_folder)
+            jax.config.update("jax_compilation_cache_dir", folder_path_jax)
+            jax.config.update("jax_persistent_cache_min_entry_size_bytes", -1)
+            jax.config.update("jax_persistent_cache_min_compile_time_secs", 0)
```
Comment (shtopane) [inline econpizza/config.py:38]: New insight - `jax_persistent_cache_min_entry_size_bytes` should be left to the default value of 0. This ensures that the cache folder won't grow indefinitely. I'll push a change for this.

**Comment by gboehl** [comment]
Are you going to open the two PRs in the original EP repository?

**Comment by shtopane** [comment]
> Are you going to open the two PRs in the original EP repository?

Of course, sorry for the delay!

**Pull request merged**